### PR TITLE
Convert drivers to use I2CDevice interface

### PR DESCRIPTION
**Pull request by lgywata**
Replaces PRs: #3904, #3905, #3910, #3933 

Tested commits: 
- AP_HAL_Linux: RCOutput_PCA9685: use I2CDevice interface
- AP_Notify: Display_SSD1306_I2C: use I2CDevice interface

### CODE REVIEW

File: libraries/AP_Notify/ToshibaLED_I2C.cpp [hunk 1]
```diff
@@ -29,49 +29,33 @@ extern const AP_HAL::HAL& hal;
...
-
+    _dev = hal.i2c_mgr->get_device(0, TOSHIBA_LED_ADDRESS);
     // take i2c bus sempahore
-    if (!i2c_sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
         return false;
     }
 
-    // disable recording of i2c lockup errors
-    hal.i2c->ignore_errors(true);
```
Comment (ricardotk) [inline libraries/AP_Notify/ToshibaLED_I2C.cpp]: Is it safe to remove this disablement of i2c lockup errors?

File: libraries/AP_ADC/AP_ADC_ADS1115.cpp [hunk 2]
```diff
@@ -1,4 +1,5 @@
 #include <AP_HAL/AP_HAL.h>
+#include <endian.h>
```
Comment (ricardotk) [inline libraries/AP_ADC/AP_ADC_ADS1115.cpp]: maybe use `#include <AP_HAL/utility/sparse-endian.h>` instead

File: libraries/AP_Notify/ToshibaLED_I2C.h [hunk 3]
```diff
@@ -17,9 +17,13 @@
 #pragma once
 
 #include "ToshibaLED.h"
+#include <AP_HAL/I2CDevice.h>
```
Comment (ricardotk) [inline libraries/AP_Notify/ToshibaLED_I2C.h]: I would put `<thistypeofincludes>` before `"thistypeofincludes"`

File: libraries/AP_Notify/ToshibaLED_I2C.cpp [hunk 4]
```diff
@@ -29,49 +29,33 @@ extern const AP_HAL::HAL& hal;
 
 bool ToshibaLED_I2C::hw_init()
 {
-    // get pointer to i2c bus semaphore
-    AP_HAL::Semaphore* i2c_sem = hal.i2c->get_semaphore();
-
+    _dev = hal.i2c_mgr->get_device(0, TOSHIBA_LED_ADDRESS);
```
Comment (ricardotk) [inline libraries/AP_Notify/ToshibaLED_I2C.cpp]: You could create a public constructor for this class like you did in some other files to put this `_dev = hal.i2c_mgr->get_device(0, TOSHIBA_LED_ADDRESS);`

File: libraries/AP_Notify/Display_SSD1306_I2C.h [hunk 5]
```diff
@@ -10,11 +11,13 @@
...
 public:
+    Display_SSD1306_I2C();
     virtual bool hw_init();
     virtual bool hw_update();
     virtual bool set_pixel(uint16_t x, uint16_t y);
     virtual bool clear_pixel(uint16_t x, uint16_t y);
 
 private:
     uint8_t _displaybuffer[SSD1306_ROWS * SSD1306_COLUMNS_PER_PAGE];
+    AP_HAL::OwnPtr<AP_HAL::I2CDevice> _dev;
```
Comment (ricardotk) [inline libraries/AP_Notify/Display_SSD1306_I2C.h]: I prefer to put object types before raw types (e.g. uint8_t, char, bool) in the declaration.
This comments goes for the other files as well.

File: libraries/AP_HAL_Linux/RCOutput_PCA9685.cpp [hunk 6]
```diff
@@ -54,20 +55,19 @@ using namespace Linux;
...
-                                             int16_t oe_pin_number) :
-    _i2c_sem(NULL),
+RCOutput_PCA9685::RCOutput_PCA9685(AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                   bool external_clock,
+                                   uint8_t channel_offset,
+                                   int16_t oe_pin_number) :
     _enable_pin(NULL),
     _frequency(50),
     _pulses_buffer(new uint16_t[PWM_CHAN_COUNT - channel_offset]),
-    _addr(addr),
```
Comment (ricardotk) [inline libraries/AP_HAL_Linux/RCOutput_PCA9685.cpp]: I would add `_dev(std::move(dev)` in this line since it follows the parameters order.

File: libraries/AP_RangeFinder/RangeFinder.cpp [hunk 7]
```diff
@@ -494,11 +494,8 @@ void RangeFinder::detect_instance(uint8_t instance)
             hal.i2c_mgr->get_device(0, AP_RANGEFINDER_PULSEDLIGHTLRF_ADDR)));
     } 
     if (type == RangeFinder_TYPE_MBI2C) {
-        if (AP_RangeFinder_MaxsonarI2CXL::detect(*this, instance)) {
-            state[instance].instance = instance;
-            drivers[instance] = new AP_RangeFinder_MaxsonarI2CXL(*this, instance, state[instance]);
-            return;
-        }
+        add_backend(AP_RangeFinder_MaxsonarI2CXL::detect(*this, instance, state[instance],
+            hal.i2c_mgr->get_device(0, AP_RANGE_FINDER_MAXSONARI2CXL_DEFAULT_ADDR)));
```
Comment (ricardotk) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: Is there any reason on why passing `hal.i2c_mgr->get_device(0, AP_RANGE_FINDER_MAXSONARI2CXL_DEFAULT_ADDR)` as an argument and not initializing with this value by default in `AP_RangeFinder_MaxsonarI2CXL` constructor?

File: libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarI2CXL.h [hunk 8]
```diff
@@ -17,16 +18,19 @@ class AP_RangeFinder_MaxsonarI2CXL : public AP_RangeFinder_Backend
 {
 public:
     // constructor
-    AP_RangeFinder_MaxsonarI2CXL(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state);
+    AP_RangeFinder_MaxsonarI2CXL(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state,
+                                 AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev);
```
Comment (ricardotk) [inline libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarI2CXL.h]: shouldn't this constructor be `private`?

File: libraries/AP_RangeFinder/RangeFinder.cpp [hunk 9]
```diff
@@ -501,10 +501,9 @@ void RangeFinder::detect_instance(uint8_t instance)
         }
     }
     if (type == RangeFinder_TYPE_LWI2C) {
-        if (AP_RangeFinder_LightWareI2C::detect(*this, instance)) {
-            state[instance].instance = instance;
-            drivers[instance] = new AP_RangeFinder_LightWareI2C(*this, instance, state[instance]);
-            return;
+        if (_address[instance]) {
+                add_backend(AP_RangeFinder_LightWareI2C::detect(*this, instance, state[instance],
+                    hal.i2c_mgr->get_device(0, _address[instance])));
```
Comment (ricardotk) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: too much spaces here

File: libraries/AP_HAL/Device.h [hunk 10]
```diff
@@ -64,6 +64,11 @@ class AP_HAL::Device {
         return transfer(buf, sizeof(buf), nullptr, 0);
     }
 
+    bool read(uint8_t *recv, uint32_t recv_len)
+    {
+        return transfer(nullptr, 0, recv, recv_len);
+    }
+
```
Comment (ricardotk) [inline libraries/AP_HAL/Device.h]: maybe this should be in another commit

File: libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.h [hunk 11]
```diff
@@ -33,17 +34,18 @@ class AP_RangeFinder_PulsedLightLRF : public AP_RangeFinder_Backend
 
 public:
     // constructor
-    AP_RangeFinder_PulsedLightLRF(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state);
+    AP_RangeFinder_PulsedLightLRF(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev);
```
Comment (ricardotk) [inline libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.h]: should't this be `private`?

File: libraries/AP_Notify/NavioLED_I2C.cpp [hunk 12]
```diff
@@ -26,37 +26,27 @@ extern const AP_HAL::HAL& hal;
 
 bool NavioLED_I2C::hw_init()
 {
-    // get pointer to i2c bus semaphore
-    AP_HAL::Semaphore* i2c_sem = hal.i2c->get_semaphore();
+    _dev = hal.i2c_mgr->get_device(1, PCA9685_ADDRESS);
```
Comment (ricardotk) [inline libraries/AP_Notify/NavioLED_I2C.cpp]: you could create a constructor for this

File: libraries/AP_Notify/NavioLED_I2C.cpp [hunk 13]
```diff
@@ -26,37 +26,27 @@ extern const AP_HAL::HAL& hal;
...
 
     // take i2c bus sempahore
-    if (!i2c_sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
         return false;
     }
 
-    // disable recording of i2c lockup errors
-    hal.i2c->ignore_errors(true);
-
```
Comment (ricardotk) [inline libraries/AP_Notify/NavioLED_I2C.cpp]: Is it safe to remove this disabling of i2c lockup errors?

File: libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.h (hunk 11 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.h]: Yes!

**Comment by ricardotk** [comment]
There are constructors that you can declare on the `header` file, this applies for all the commits that you add a constructor that only sets the class attributes

File: libraries/AP_Notify/NavioLED_I2C.cpp (hunk 13 above)
Comment (lgywata) [inline libraries/AP_Notify/NavioLED_I2C.cpp]: Yes. Thanks.

File: libraries/AP_RangeFinder/RangeFinder.cpp [hunk 14]
```diff
@@ -461,6 +461,18 @@ void RangeFinder::update(void)
...
+    }
+    if (num_instances == RANGEFINDER_MAX_INSTANCES) {
+        AP_HAL::panic("Too many RANGERS backends");
+    }
+
+    drivers[num_instances++] = driver;
+}
     
 /*
   detect if an instance of a rangefinder is connected.
```
Comment (ricardotk) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: could you add a better commit message here? e.g. why is this `add_backend` needed now?

File: libraries/AP_ADC/AP_ADC_ADS1115.cpp (hunk 2 above)
Comment (lgywata) [inline libraries/AP_ADC/AP_ADC_ADS1115.cpp]: ok.

File: libraries/AP_Notify/Display_SSD1306_I2C.h (hunk 5 above)
Comment (lgywata) [inline libraries/AP_Notify/Display_SSD1306_I2C.h]: Ack.

File: libraries/AP_Notify/NavioLED_I2C.cpp (hunk 12 above)
Comment (lgywata) [inline libraries/AP_Notify/NavioLED_I2C.cpp]: ack.

File: libraries/AP_HAL/Device.h (hunk 10 above)
Comment (lgywata) [inline libraries/AP_HAL/Device.h]: Fair.

File: libraries/AP_RangeFinder/RangeFinder.cpp (hunk 9 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: Ack.

File: libraries/AP_RangeFinder/RangeFinder.cpp (hunk 7 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: No reason other than inertia. :p I'll change it, thanks.

File: libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarI2CXL.h (hunk 8 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarI2CXL.h]: Yes.

**Comment by lgywata** [comment]
see #4222
